### DISCUSSION TITLE
next-img – a plugin for embedding optimized images

### DISCUSSION THREAD

**Discussion by KidkArolis** [vercel/next.js#13947] [Show and tell]
Hi 👋,

I've been working on and off over the past month on a plugin for making it easier to add images to Next.js projects.

Adding images to websites these days is a little on the tricky side. You need to take into account the performance, different screen sizes, different screen densities, modern image formats, and size it all just right to fit into the specific designs. I wanted to automate this better.

`next-img` is a:

1. Next.js plugin (plugs into Next.js build pipeline)
2. Webpack loader (does all the image transformation)
3. React component (abstracts the complex markup)

all in one.

You just point it at an image, specify required sizes and it will produce:

1. an image for each breakpoint
2. an image for each screen density
3. original format and webp for smaller file size
4. and output all of the right markup

For example, it will turn:

```js
<Picture src={require('../images/coffee1.jpg?sizes=375,860')} />
```

into:

```js
<picture>
    <source
        type="image/webp"
        srcSet="/next-img/_next/static/images/coffee1-375@1x-d04d19c08afad1ce95e7f102961c1a7e.webp 375w,
                /next-img/_next/static/images/coffee1-375@2x-9d17025a66d743b5ba03e9046e9f4365.webp 750w,
                /next-img/_next/static/images/coffee1-860@1x-b811b4e74128a685ea8029bcde250d74.webp 860w,
                /next-img/_next/static/images/coffee1-860@2x-71e3c8575537ad7bdcacee55fa91268c.webp 1720w"
        sizes="(max-width: 768px) 375px, 860px"
    />
    <source
        type="image/jpeg"
        srcSet="/next-img/_next/static/images/coffee1-375@1x-e2b9b524b9126743aa9764cfd425fa4e.jpg 375w,
                /next-img/_next/static/images/coffee1-375@2x-0a1620261148881f98ebdc78ed4ec085.jpg 750w,
                /next-img/_next/static/images/coffee1-860@1x-da2a8aa62cc17fe9e7ca51461981387c.jpg 860w,
                /next-img/_next/static/images/coffee1-860@2x-11753d1994d0b6d94afa566e9b3cd4ea.jpg 1720w"
        sizes="(max-width: 768px) 375px, 860px"
    />

    <img
        src="/next-img/_next/static/images/coffee1-375@1x-e2b9b524b9126743aa9764cfd425fa4e.jpg"
        srcSet="/next-img/_next/static/images/coffee1-375@1x-e2b9b524b9126743aa9764cfd425fa4e.jpg 375w,
                /next-img/_next/static/images/coffee1-375@2x-0a1620261148881f98ebdc78ed4ec085.jpg 750w,
                /next-img/_next/static/images/coffee1-860@1x-da2a8aa62cc17fe9e7ca51461981387c.jpg 860w,
                /next-img/_next/static/images/coffee1-860@2x-11753d1994d0b6d94afa566e9b3cd4ea.jpg 1720w"
    />
</picture>
```

One of my favorite aspects of this plugin is the **persistent caching**. Image processing is slow. Doing that on every `next` startup or `next build` is not ideal. Therefore, `next-img` caches all of the processed images in the git repo to avoid any reprocessing when developing / deploying the project in the future.

Docs in the repo: https://github.com/humaans/next-img/
Examples: https://humaans.github.io/next-img/

I hope you find this useful 🙏

**Comment by franknoel** [comment]
That's great, I'll definetly take a look!
The main issue that I have with your approach, and the one of `next-optimized-images`, is that it only support local images. Any project with a headless architecture for instance is unable to benefit from your plugin.
